src/registry.rs

```diff
@@ -83,47 +124,47 @@ impl Registry {
         self.r.write().unwrap().unregister(c)
     }
 
-    pub fn write_pb<T: Write>(&self, _: &mut T) -> Result<()> {
-        Ok(())
+    fn gather(&self) -> Vec<proto::MetricFamily> {
+        self.r.read().unwrap().gather()
     }
 
-    pub fn write_text<T: Write>(&self, _: &mut T) -> Result<()> {
-        Ok(())
-    }
-}
+    pub fn scrap(&self, writer: &mut Write, encoder: &Encoder) -> Result<usize> {
```

Comment:
you can remove scrap now. outer uses endoder + gather is ok.
Btw, why using writer + encoder at same time? IMO, encoder should wrap writer and do the writer internal.